library/alloc/src/sync/tests.rs

```diff
@@ -101,6 +101,45 @@ fn try_unwrap() {
     assert_eq!(Arc::try_unwrap(x), Ok(5));
 }
 
+#[test]
+fn unwrap_or_drop() {
+    // FIXME: Is doing this kind of loop reasonable? I tested `Arc::try_unwrap(x).ok`
+    // and it makes this kind of assersion fail in roughly every second run somewhere
+    // between 1000 and 5000 iterations; I feel like doing a sinlge iteration is too
+    // unlikely to catch anything interesting but doing too many is way too slow
+    // for a test that wouldn't ever fail for any reasonable implementation
+
+    for _ in 0..100
+    // ^ increase chances of hitting uncommon race conditions
+    {
+        use std::sync::Arc;
+        let x = Arc::new(3);
+        let y = x.clone();
+        let r_thread = std::thread::spawn(|| Arc::try_unwrap(x).ok());
+        let s_thread = std::thread::spawn(|| Arc::try_unwrap(y).ok());
+        let r = r_thread.join().expect("r_thread panicked");
+        let s = s_thread.join().expect("s_thread panicked");
+        assert!(
+            matches!((r, s), (None, Some(3)) | (Some(3), None)),
+            "assertion failed: unexpected result `{:?}`\
+            \n  expected `(None, Some(3))` or `(Some(3), None)`",
+            (r, s),
+        );
+    }
+
+    let x = Arc::new(3);
+    assert_eq!(Arc::unwrap_or_drop(x), Some(3));
+
+    let x = Arc::new(4);
+    let y = x.clone();
```

Comment:
Here too

Reply:
This one is copied from the `try_unwrap` test above.

Reply:
I should update it everywhere. I'll get to it one day 🤞.

You can still change it here to promote `Arc::clone` a little more. :)

Reply:
I should update it everywhere. I'll get to it one day 🤞.

You can still change it here to promote `Arc::clone` a little more. :)

Reply:
I should update it everywhere. I'll get to it one day 🤞.

You can still change it here to promote `Arc::clone` a little more. :)

Reply:
Well, one distinction to be made is between the doc-test where the source code will be visible in the documentation vs. the ordinary test for which mostly the functionality matters. Anyways, I’ll apply your suggestions to keep consistency within my test function and with the doctest.

Also, are there really 3 copies of your answer?

Reply:
Yeah, I shouldn't review while on the train, Github doesn't like loosing network and getting it back every 5 seconds